src/lambda_toolshed/papillon.cljc

```diff
@@ -1,30 +1,34 @@
 (ns lambda-toolshed.papillon
-  (:require
-   [clojure.core.async :refer [<! go go-loop take! put! chan]]
-   [clojure.core.async.impl.protocols :refer [ReadPort]]
-   [lambda-toolshed.papillon.async]
-   #?@(:cljs ([goog.string :as gstring]
-              goog.string.format))))
+  (:require [lambda-toolshed.papillon.protocols :refer [promote emerge]]
+            #?@(:cljs ([goog.string :as gstring]
+                       goog.string.format))))
 
 (def ^:private fmt
   #?(:clj format :cljs gstring/format))
 
-(defprotocol InterceptorPromotable
-  (promote [this] "Promote this to an interceptor"))
-
-(extend-protocol InterceptorPromotable
-  #?(:clj clojure.lang.Var :cljs cljs.core/Var) (promote [this] (deref this))
-  #?@(:clj (clojure.lang.IPersistentMap (promote [this] this))
-      :cljs (cljs.core/PersistentArrayMap (promote [this] this)
-                                          cljs.core/PersistentHashMap (promote [this] this)
-                                          cljs.core/PersistentTreeMap (promote [this] this)))
-  ;; TODO: extend the default in cljs
-  #?(:clj Object :cljs object) (promote [this] (throw (ex-info "Not promoteable to an Interceptor!"
-                                                               {:obj this :type (type this)}))))
+(defn- identify
+  "Return the string identifier of the given interceptor `ix`."
+  [ix]
+  (or (:name ix) (str "ix-" (hash ix))))
+
 (defn enqueue
   [ctx ixs]
   (update ctx ::queue into (map promote) ixs))
 
+(defn initialize
+  "Inialize the given context `ctx` with the necessary data structures to process
+   the interceptor chain `ixs`.  Note: starting with an initial context that
+   contains the key `:lambda-toolshed.papillon/error` is undefined and may change."
+  ([ixs] (initialize ixs {}))
+  ([ixs ctx]
+   (-> ctx
+       (vary-meta assoc :type ::ctx)
+       (assoc ::queue #?(:clj clojure.lang.PersistentQueue/EMPTY
+                         :cljs cljs.core/PersistentQueue.EMPTY))
+       (assoc ::stack [])
+       (assoc ::stage :enter)
```

Comment:
are there benefits I'm not remembering to separating all of these `assoc` calls out? or would the more succinct single call to `assoc` with multiple kv pairs work?

```clojure
(assoc ::queue #?(:clj clojure.lang.PersistentQueue/EMPTY
                         :cljs cljs.core/PersistentQueue.EMPTY)
       ::stack []
       ::stage :enter)
```

Reply:
Zero benefit, AFAIK.  This is a sign of evolving instead of creating.  Fix incoming.

Reply:
ok, cool. that makes sense, sounds good.